h2o-algos/src/main/java/hex/ensemble/StackedEnsembleModel.java

```diff
@@ -116,24 +120,40 @@ public void initMetalearnerParams(Metalearner.Algorithm algo) {
   protected Frame predictScoreImpl(Frame fr, Frame adaptFrm, String destination_key, Job j, boolean computeMetrics, CFuncRef customMetricFunc) {
     Frame levelOneFrame = new Frame(Key.<Frame>make("preds_levelone_" + this._key.toString() + fr._key));
 
-    // TODO: we should be able to parallelize scoring of base models
+    List<H2O.H2OCountedCompleter> tasks = new ArrayList<>();
+    final String seKey = this._key.toString();
+
+    // Run scoring of base models in parallel
     for (Key<Model> baseModelKey : _parms._base_models) {
       if (isUsefulBaseModel(baseModelKey)) {
-        Model baseModel = baseModelKey.get();
-
-        Frame basePreds = baseModel.score(
-                fr,
-                "preds_base_" + this._key.toString() + fr._key,
-                j,
-                false
-        );
-
-        StackedEnsemble.addModelPredictionsToLevelOneFrame(baseModel, basePreds, levelOneFrame);
-        DKV.remove(basePreds._key); //Cleanup
-        Frame.deleteTempFrameAndItsNonSharedVecs(basePreds, levelOneFrame);
+        tasks.add(
+                H2O.submitTask(
+                        new H2O.H2OCountedCompleter() {
+                          @Override
+                          public void compute2() {
+                            Model baseModel = baseModelKey.get();
+
+                            Frame basePreds = baseModel.score(
+                                    fr,
+                                    "preds_base_" + seKey + fr._key,
+                                    j,
+                                    false
+                            );
+
+                            StackedEnsemble.addModelPredictionsToLevelOneFrame(baseModel, basePreds, levelOneFrame);
+                            DKV.remove(basePreds._key); //Cleanup
+                            Frame.deleteTempFrameAndItsNonSharedVecs(basePreds, levelOneFrame);
+                            tryComplete();
+                          }
+                        }));
```

Comment:
I did some implementation using the MRTask but MRTask along with H2O's multi-node rpc/DKV are still new to me so I have some questions:

- Assumption: `Frame fr` is a lightweight object that contains references to `DKV` to its `Vec`s; `Vec`s contain `Chunk`s and `Chunk`s can be distributed on multiple nodes. 
  - When I run from local node `model.score(fr)` will it download the parts of the frame to local node and score it locally or does it broadcast the model to the nodes and evaluate the model in parallel for each part of the frame?
    - (The former isn't very practical (every node fetching the same data for the distributed base model scoring) so I am assuming that the latter is closer to the actual implementation.)
    - If the latter is true, would it really make a significant difference if I would score from single node or mutliple nodes? 

- If I run MR on one element, it seems to me that it won't run the reduce step. Is that correct? If so, I should "clean the dirt" in the map phase (current implementation).

Reply:
sorry, I missed that comment:
> If I run MR on one element, it seems to me that it won't run the reduce step. Is that correct? If so, I should "clean the dirt" in the map phase (current implementation).

I think you're right. But the dirt you're creating is specific to the map phase so it makes sense to clean it there.

> When I run from local node model.score(fr) will it download the parts of the frame to local node and score it locally or does it broadcast the model to the nodes and evaluate the model in parallel for each part of the frame?

model scoring is implemented through the `BigScore` task which is an `MRTask`, so it has to be the latter.

> If the latter is true, would it really make a significant difference if I would score from single node or mutliple nodes?
very good point indeed... probably not...
Sorry to reply after your did the changes, but do you see any difference?

Reply:
I didn't test it well enough (just ~5 runs) to say if there is some difference - they were very close sometimes the first was faster sometimes the second solution was faster.

Reply:
> If the latter is true, would it really make a significant difference if I would score from single node or mutliple nodes?

I think using a single node shouldn't be slower, maybe by just a little bit. There is an advantage of using a single node (in general):
1. are logs are in one place - not super important but can be convenient
2. model metrics are stable - eg. AUC will be slightly different depending on the node that initiates the scoring (the order of the reduces will be different and the histogram bins might be different as a result). In this case, you don't care about the metrics.
3. you can better control parallelism.

I would go for `LocalMR` implementation of the scoring, it will give you a better option of controlling how many parallel scoring tasks you want to invoke.

Reply:
thanks @michalkurka we also agreed on reverting back to local scoring